Avoid mutex ordering issues.

bbockelm opened the pull request:
Prevent m_source_mutex from being held when OpenHandler::open is called
(which will then take the internal m_mutex).  This avoids a lock re-ordering
issue.

bbockelm committed:
Avoid mutex ordering issues.

Prevent m_source_mutex from being held when OpenHandler::open is called
(which will then take the internal m_mutex).  This avoids a lock re-ordering
issue.

cmsbuild commented:
A new Pull Request was created by  @bbockelm (Brian Bockelman) for CMSSW_7_5_X.

It involves the following packages:

Utilities/XrdAdaptor


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

Following commands in first line of a comment are recognized

* **+1|approve[d]|sign[ed]**: L1/L2's to approve it
* **-1|reject[ed]**: L1/L2's to reject it
* **assign &lt;category&gt;[,&lt;category&gt;[,...]]**: L1/L2's to request signatures from other categories
* **unassign &lt;category&gt;[,&lt;category&gt;[,...]]**: L1/L2's to remove signatures from other categories
* **hold**: L1/all L2's/release manager to mark it as on hold
* **unhold**: L1/user who put this PR on hold
* **merge**: L1/release managers to merge this request
* **[@cmsbuild,] please test**: L1/L2 and selected users to start jenkins tests
* **[@cmsbuild,] please test with cms-sw/cmsdist#&lt;PR&gt;**: L1/L2 and selected users to start jenkins tests using externals from cmsdist

slava77 commented:
@bbockelm how serious is this one?
Is it worth to make another release or just wait until other things accumulate in the next few days.

bbockelm commented:
It can likely wait.  @Dr15Jones noticed this in the relvals - we see no reason why it couldn't trigger in the T0, but the likelihood of hitting the race is smaller than the other XrdAdaptor fix.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: a9923b0a1ed03ffbc6465a50577b00b6562b08e4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12613/10029/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12613/10029/summary.html

slava77 commented:
@bbockelm 
we will also need 76X/80X PR(s)

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdRequestManager.cc:
```diff
@@ -370,10 +370,14 @@ RequestManager::checkSourcesImpl(timespec &now, IOSize requestSize)
-    m_open_handler->open();
     m_lastSourceCheck = now;
+    sentry.unlock();
+    m_open_handler->open();
```
need to reacquire the lock after this line.

bbockelm committed:
Use RAII to reacquire locks.

To avoid having to explicitly re-acquire locks (and worry about
exception safety), use a small RAII trick with shared_ptr to
automatically reacquire things.

cmsbuild commented:
Pull request #12613 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please check and sign again.

bbockelm commented:
Fixed @Dr15Jones code comment.

I'll open PRs for the other branches once this goes in (and we can cherrypick).

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
I'll see if I can use this change on jobs which have been failing

cmsbuild commented:
+1
Tested at: 255a87558756c5caff3e3a117d5ff116e2802228
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12613/10042/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12613/10042/summary.html

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdRequestManager.cc:
```diff
@@ -370,10 +370,16 @@ RequestManager::checkSourcesImpl(timespec &now, IOSize requestSize)
+    sentry.unlock();
+    // Automatically reacquire the lock after this code block finishes.
+    std::shared_ptr<const void> reacquire_lock_sentry(nullptr, [&sentry](void*) {sentry.lock();});
+    m_open_handler->open();
```
Turns out this is not safe. There is a comment in OpenHandler::open() which explicitly states that this function isn't thread safe and requires a lock on it to guarantee you can't call it from multiple threads. The problem is an unguarded access to a member data.

Dr15Jones commented:
@bbockelm What's next try for this fix?

Dr15Jones commented:
@bbockelm Should we go with just one mutex?

bbockelm commented:
Nope - it turns out the fix is more difficult than that.

Basically, depending on the context, the callback handler may-or-may-not be invoked with a Xrootd mutex held (I really, really hate this fact).  This same mutex will be needed for XrdFile::open.

So, any mutex held within the callback MUST NOT be held when we call XrdFile::open.  This is why we unlocked `m_mutex` before; we don't currently have this problem with `m_source_mutex` because of the deadlock that triggered this ticket.  If we merge the two together, we'll bring back the old issue.

Rather, I think we need to get rid of the use of `m_source_mutex` in the callback.  It is protecting read-only data that changes whenever a server throws an error - so, on the order of minutes.  Further, it is only used to format error messages.

I am working to pre-format the parts of the error message that require `m_source_mutex` and caching them in an lock-free manner.  This removes the use of `m_source_mutex` in the callback function, allowing us to keep it held when we call `XrdFile::open`.

davidlange6 commented:
Hi - I'm cleaning up the 75x CMSSW release queue, so closing issues not directly related to analysis or finalizing the HI MC production. Please make sure these requests are included in the 80x release already. Thanks!
